netengine/backends/snmp/openwrt.py

```diff
@@ -349,60 +339,196 @@ def interfaces_to_dict(self):
+            offset = datetime.timedelta(hours=hours_from_utc, minutes=minutes_from_utc)
+            if direction == b'-':
+                offset = -offset
+            return int(
+                datetime.datetime(
+                    year,
+                    month,
+                    day,
+                    hour,
+                    minutes,
+                    seconds,
+                    deci_seconds * 100_000,
+                    tzinfo=pytz.utc,
+                ).timestamp()
+            )
+
     @property
     def RAM_total(self):
         """
         returns the total RAM of the device
         """
         return int(self.get_value('1.3.6.1.2.1.25.2.3.1.5.1'))
 
+    @property
+    def RAM_shared(self):
+        """
+        returns the shared RAM of the device
+        """
+        return int(self.get_value('1.3.6.1.2.1.25.2.3.1.6.8'))
+
+    @property
+    def RAM_cached(self):
+        """
+        returns the cached RAM of the device
+        """
+        return int(self.get_value('1.3.6.1.2.1.25.2.3.1.5.7'))
+
+    @property
+    def RAM_free(self):
+        """
+        returns the free RAM of the device
+        """
+        RAM_used = int(self.get_value('1.3.6.1.2.1.25.2.3.1.6.1'))
+        RAM_free = self.RAM_total - (self.RAM_cached + RAM_used)
+        return RAM_free
+
+    @property
+    def SWAP_total(self):
+        """
+        returns the total SWAP of the device
+        """
+        return int(self.get_value('1.3.6.1.2.1.25.2.3.1.5.10'))
+
+    @property
+    def SWAP_free(self):
+        """
+        returns the free SWAP of the device
+        """
+        SWAP_used = int(self.get_value('1.3.6.1.2.1.25.2.3.1.6.10'))
+        SWAP_free = self.SWAP_total - SWAP_used
+        return SWAP_free
+
+    @property
+    def CPU_count(self):
+        """
+        returns the count of CPUs of the device
+        """
+        return len(self.next('1.3.6.1.2.1.25.3.3.1.2')[3])
+
+    @property
+    def resources_to_dict(self):
+        """
+        returns an ordered dict with hardware resources information
+        """
+        result = self._dict(
+            {
+                'cpus': self.CPU_count,
+                'memory': {
+                    'total': self.RAM_total,
+                    'shared': self.RAM_shared,
+                    'free': self.RAM_free,
+                    'cached': self.RAM_cached,
+                },
+                'swap': {'total': self.SWAP_total, 'free': self.SWAP_free},
+            }
+        )
+        return result
+
+    @property
+    def neighbors(self):
+        """
+        returns a dict with neighbors information
+        """
+        states_map = {
+            '1': 'REACHABLE',
+            '2': 'STALE',
+            '3': 'DELAY',
+            '4': 'PROBE',
+            '5': 'INVALID',
+            '6': 'UNKNOWN',
+            '7': 'INCOMPLETE',
+        }
+
+        # TODO: find a way to extract IP address from the OID
```

Comment:
I spent some time on this and this is what i found for OpenWrt device. 

Following is output for my laptop as neighbour of the router. I replaced my MAC with `FF FF FF FF FF FF`.

```
$ snmpwalk -v2c -c public  192.168.1.1 1.3.6.1.2.1.4.35.1.4
iso.3.6.1.2.1.4.35.1.4.5.1.4.192.168.1.136 = Hex-STRING: FF FF FF FF FF FF
iso.3.6.1.2.1.4.35.1.4.5.2.16.254.128.0.0.0.0.0.0.35.113.163.173.241.78.112.113 = Hex-STRING: FF FF FF FF FF FF
iso.3.6.1.2.1.4.35.1.4.5.2.16.254.128.0.0.0.0.0.0.77.92.143.78.43.2.155.123 = Hex-STRING: FF FF FF FF FF FF
```

You need to pay attention for the OID here. For IPv4 addresses the 12th index is `4` and for IPv6 address 12th index is `16`. 

The IPv6 addresses are represented in decimal octets here. You will have to convert them to hexadecimals. Check if there's an existing library which can do this to make our work simple. 

Check this pseudocode for example, improve it wherever required:

```
for neighbor in neighbors:
    oid = neighbour[0][0].getOid()
    if oid[12] == 4:
        ip = oid[13:]
    else:
        ip = convert_decimal_octets_to_hexademical_blocks(oid[13:])
```

Reply:
Done :+1: